lib/dialogue-agent/conversation.ts

```diff
@@ -548,4 +570,48 @@ export default class Conversation extends events.EventEmitter {
             console.log('Almond sends link: '+ url);
         return this._addMessage({ type: MessageType.LINK, url, title });
     }
+
+    appendLog(turn : DialogueTurn) {
+        this._log.push(turn);
+    }
+
+    upvoteLast() {
+        if (this._log.length === 0)
+            throw new Error('No response found to be rated');
+        const last = this._log[this._log.length - 1];
+        last.rate = '👍';
+    }
+
+    downvoteLast() {
+        if (this._log.length === 0)
+            throw new Error('No response found to be rated');
+        const last = this._log[this._log.length - 1];
+        last.rate = '👎';
+    }
+
+    commentLast(comment : string) {
+        if (this._log.length === 0)
+            throw new Error('No response found to be commented');
+        const last = this._log[this._log.length - 1];
+        last.comment = comment;
+    }
+
+    async saveLog() {
+        const dir = this._engine.platform.getWritableDir();
+        const logfile = path.join(dir, this.id + '.txt');
+        const serializer = new DialogueSerializer({ annotations: true });
+
+        const output = fs.createWriteStream(logfile);
+
+        serializer.pipe(output);
+        serializer.write({ id: this.id, turns : this._log });
+        serializer.end();
+
+        await StreamUtils.waitFinish(output);
+    }
+
+    loadLog() {
+        const dir = this._engine.platform.getWritableDir();
+        return fs.readFileSync(path.join(dir, this.id + '.txt')).toString();
```

Comment:
Is this for download? Maybe return a stream or filename instead of a blob?

Reply:
This was actually just used for testing... I'm actually not sure how the download from almond-cloud would be implemented. Any existing code I can refer to?

Reply:
Download from almond-server is easy: return a filename, and just stream the file content in the download API with `fs.createReadStream(filename).pipe(res)`. From almond-cloud is tricky because the frontend processes might not have access to the same filesystem where the backend process resides. You would need some sort of streaming protocol.